keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/constraint/BakeConstraintEvaluator.kt

```diff
@@ -0,0 +1,62 @@
+package com.netflix.spinnaker.keel.bakery.constraint
+
+import com.netflix.frigga.ami.AppVersion
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.Locatable
+import com.netflix.spinnaker.keel.api.RegionSpec
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.bakery.api.BakeConstraint
+import com.netflix.spinnaker.keel.clouddriver.ImageService
+import com.netflix.spinnaker.keel.constraints.ConstraintEvaluator
+import com.netflix.spinnaker.keel.constraints.SupportedConstraintType
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+
+class BakeConstraintEvaluator(
+  override val eventPublisher: ApplicationEventPublisher,
+  private val imageService: ImageService,
+  private val dynamicConfigService: DynamicConfigService
+) : ConstraintEvaluator<BakeConstraint> {
+  override val supportedType = SupportedConstraintType<BakeConstraint>("bake")
+  override fun canPromote(
+    artifact: DeliveryArtifact,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    targetEnvironment: Environment
+  ): Boolean {
+    val locatableResources = targetEnvironment
+      .resources
+      .map(Resource<*>::spec)
+      .filterIsInstance<Locatable<*>>()
+
+    val regions: List<String> = locatableResources.flatMap {
+      it.locations.regions.map(RegionSpec::name)
+    }
+
+    return if (regions.isEmpty()) {
+      true
+    } else {
+      log.debug("Searching for baked image for %s in %s", version, regions.joinToString())
+
+      runBlocking {
+        imageService.getLatestNamedImageWithAllRegionsForAppVersion(
+          AppVersion.parseName(version),
+          defaultImageAccount,
+          regions
+        ) != null
+      }
+    }
+  }
+
+  private val defaultImageAccount: String
+    get() = dynamicConfigService.getConfig("images.default-account", "test")
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private inline fun <reified T> DynamicConfigService.getConfig(configName: String, defaultValue: T) =
```

Comment:
This seems like it would be useful elsewhere. Move to a utility class somewhere?

Reply:
Yeah, I copied it from another place in our codebase. Would be good to remove that duplication